# **Tugas Individu 1**
#### Nama = Akmal Kafli Anan
#### NIM  = 24060124120042
#### Kelas  = A

In [26]:
# Libary
import numpy as np
import sympy as sp
from tabulate import tabulate

## *Deret Taylor*
Deret Taylor dari sebuah fungsi riil atau fungsi kompleks f(x) yang terdiferensialkan takhingga dalam sebuah pemetaan sebuah bilangan riil atau kompleks a adalah deret pangkat <br>
$\displaystyle f(x)=f(a)+{\frac {f'(a)}{1!}}(x-a)+{\frac {f''(a)}{2!}}(x-a)^{2}+{\frac {f^{(3)}(a)}{3!}}(x-a)^{3}+\cdots $ <br>
yang dalam bentuk lebih ringkas dapat dituliskan sebagai <br>
$\displaystyle f(x)=\sum _{n=0}^{\infty }{\frac {f^{(n)}(a)}{n!}}\,(x-a)^{n}$ <br>

In [20]:
def turunan(f,a,n):
    '''
    Fungsi untuk mencari turunan dari f

    parameters:
    f(function) = fungsi yang mau diturunkan
    a(float) = nilai dari x
    n(int) = ordo turunan
    '''
    
    x = sp.symbols('x')                                 # mendefinisikan variabel simbolik
    turunan = sp.diff(f(x), x, n)                       # menghitung turunan ke -n
    f_prime = sp.lambdify(x, turunan, 'numpy')          # konvensi ko fungsi lambda
    return f_prime(a)

def DeretTaylor(f, x, a, n):
    '''
    Fungsi untuk mencari nilai x dari deret taylor

    parameters:
    f(function) = fungsi yang mau diturunkan
    x(float) = nilai yang mau di cari
    a(float) = titik pusat
    n(int) = banyak iterasi
    '''
    
    hasil = 0                          # deklarasi varible untuk hasil perulangan
    data = []                          # deklarasi list untuk data tabulate
    headers = ["n", "Output"]          # header dari tabulate
    
    # perulangan dari 0 - n
    for i in range(n + 1):
        derivative = turunan(f, a, i)                                 # turunan 
        hasil = hasil + ((derivative/sp.factorial(i))*(x-a)**i)       # menghitung hasil + hasil sebelumnya
        data.append([i, hasil])                                       # menambah hasil pada tabulate

    nilai_sejati = f(x)                # nilai sejati fungsi tersebut
    galat = abs(nilai_sejati-hasil)    # mencari galat sejati
    
    print(tabulate(data, headers=headers, tablefmt="pretty"))    # meng-print table dari tabulate
    print(f"Nilai hampiran = {hasil}")                           # meng-print hasil akhir dari deret taylor



### Deret taylor untuk cos(x) dengan x = 0,2, a = 1 dan n = 5

In [7]:
#Aplikasi

print("============================Untuk fungsi Cos(x)============================")
f = lambda x : sp.cos(x)
DeretTaylor(f,0.2,1,5)

============================Untuk fungsi Cos(x)============================
+---+-------------------+
| n |      Output       |
+---+-------------------+
| 0 | 0.540302305868140 |
| 1 | 1.21347909371446  |
| 2 | 1.04058235583665  |
| 3 | 0.968776831799712 |
| 4 | 0.977997991153195 |
| 5 | 0.980295767922377 |
+---+-------------------+
Nilai hampiran = 0.980295767922377


### Deret taylor untuk $\displaystyle\frac{1}{{{1}-{2x}}}$ dengan x = 0,2, a = 1 dan n = 5

In [18]:
# Aplikasi
print("============================Untuk fungsi 1/1-2x============================")
g = lambda x : 1/(1-2*x)
DeretTaylor(g,0.2,1,5)    

============================Untuk fungsi 1/1-2x============================
+---+-------------------+
| n |      Output       |
+---+-------------------+
| 0 | -1.00000000000000 |
| 1 | -2.60000000000000 |
| 2 | -5.16000000000000 |
| 3 | -9.25600000000000 |
| 4 | -15.8096000000000 |
| 5 | -26.2953600000000 |
+---+-------------------+
Nilai hampiran = -26.2953600000000


## *Deret Maclaurin*
Kasus khusus adalah bila fungsi diperluas di sekitar a = 0, maka deretnya dinamakan deret Maclaurin, yang merupakan deret Taylor baku. Kasus a = 0 paling sering muncul dalam praktek. <br>
$\displaystyle f(x)=f(0)+{\frac {f'(0)}{1!}}(x)+{\frac {f''(0)}{2!}}(x)^{2}+{\frac {f^{(3)}(0)}{3!}}(x)^{3}+\cdots $ <br>
yang dalam bentuk lebih ringkas dapat dituliskan sebagai <br>
$\displaystyle f(x)=\sum _{n=0}^{\infty }{\frac {f^{(n)}(0)}{n!}}\,(x)^{n}$ <br>

2. untuk $\displaystyle\frac{1}{{{1}-{2x}}}$<br>
    $\displaystyle\frac{1}{{{1}-{2x}}}= \sum _{n=0}^{\infty } 2x^n$

1. untuk cos(x) <br>
   $\displaystyle \cos{{x}} = \sum _{n=0}^{\infty }{\frac {(-1)^n}{(2n)!}}\,x^{2n} = $ $\displaystyle{1}-\frac{{x}^{2}}{{{2}!}}+\frac{{x}^{4}}{{{4}!}}$<br><br>

In [24]:
def Maclaurin_cos(x,n):
    '''
    Fungsi untuk mencari nilai hampiran dari fungsi cos(x)

    parameters:
    x(float) = nilai yang mau di cari
    n(int) = banyak iterasi
    '''
    
    hasil = 0                          # deklarasi varible untuk hasil perulangan
    data = []                          # deklarasi list untuk data tabulate 
    headers = ["n", "Output"]          # header dari tabulate
    # perulangan dari 0 - n
    
    for i in range(n + 1):
        hasil = hasil + ((-1)**i) * (x**(2*i)) / sp.factorial(2*i)    # menghitung hasil + hasil sebelumnya
        data.append([i, hasil])                                       # menambah hasil pada tabulate
        
    nilai_sejati = sp.cos(x)           # nilai sejati fungsi tersebut
    galat = abs(nilai_sejati-hasil)    # mencari galat sejati
                                                              
    print(tabulate(data, headers=headers, tablefmt="pretty"))     # meng-print table dari tabulate
    print(f"Nilai hampiran = {hasil}")                            # meng-print hasil akhir dari deret taylor   
    print (f"Nilai sejati = {nilai_sejati}")                      # meng-print nilai sejati
    print (f"Galat = {galat}\n")                                  # meng-print nilai sejati


# Aplikasi
print("============================Untuk fungsi Cos(x)============================")
Maclaurin_cos(0.2,5)

============================Untuk fungsi Cos(x)============================
+---+-------------------+
| n |      Output       |
+---+-------------------+
| 0 | 1.00000000000000  |
| 1 | 0.980000000000000 |
| 2 | 0.980066666666667 |
| 3 | 0.980066577777778 |
| 4 | 0.980066577841270 |
| 5 | 0.980066577841242 |
+---+-------------------+
Nilai hampiran = 0.980066577841242
Nilai sejati = 0.980066577841242
Galat = 0



In [25]:
# Fungsi untuk mencari nilai hampiran dari fungsi 1/1-2x (x = nilai, n = suku)
def Maclaurin_2x(x,n):
    '''
    Fungsi untuk mencari nilai hampiran dari fungsi 1/1-2x

    parameters:
    x(float) = nilai yang mau di cari
    n(int) = banyak iterasi
    '''
    
    hasil = 0                          # deklarasi varible untuk hasil perulangan
    data = []                          # deklarasi list untuk data tabulate 
    headers = ["n", "Output"]          # header dari tabulate
    
    # perulangan dari 0 - n
    for i in range(n + 1):
        hasil = hasil + ((2*x)**i)                                     # menghitung hasil + hasil sebelumnya
        data.append([i, hasil])                                        # menambah hasil pada tabulate

    nilai_sejati = 1/(1-2*x)           # nilai sejati fungsi tersebut
    galat = abs(nilai_sejati-hasil)    # mencari galat sejati
        
    print(tabulate(data, headers=headers, tablefmt="pretty"))     # meng-print table dari tabulate
    print(f"Nilai hampiran = {hasil}")                            # meng-print hasil akhir dari deret taylor
    print (f"Nilai sejati = {nilai_sejati}")                      # meng-print nilai sejati
    print (f"Galat = {galat}\n")                                  # meng-print nilai sejati

# Untuk fungsi 1/1-2x dengan x = 0.2, a = 0, dan n = 5
print("============================Untuk fungsi 1/1-2x============================")
Maclaurin_2x(0.2,5)

============================Untuk fungsi 1/1-2x============================
+---+--------------------+
| n |       Output       |
+---+--------------------+
| 0 |        1.0         |
| 1 |        1.4         |
| 2 |        1.56        |
| 3 |       1.624        |
| 4 | 1.6496000000000002 |
| 5 | 1.6598400000000002 |
+---+--------------------+
Nilai hampiran = 1.6598400000000002
Nilai sejati = 1.6666666666666667
Galat = 0.0068266666666665365



Dari code tersebut dapat diketahui: <br>
1. fungsi cos(x) <br>
   Galat mutlak $\displaystyle={\left|{0.980066577841242}-{0.980066577841242}\right|}={0.0000000}$ <br>
   Galat relatif $\displaystyle={\left(\frac{0.0000000}{{0.980066577841242}}\right)}\cdot{100}\%={0}\%$ <br>
   galat relatif hampiran $\displaystyle={\left(\frac{0.0000000}{{0.980066666666667}}\right)}\cdot{100}\%={0}\%$ <br>
2. fungsi 1/1-2x <br>
   Galat mutlak $\displaystyle={\left|{1.6666666666666667}-{1.6598400000000002}\right|}={0.0068266666666665365}$ <br>
   Galat relatif $\displaystyle={\left(\frac{0.0068266666666665365}{{1.6666666666666667}}\right)}\cdot{100}\%={0.4096}\%$ <br>
   galat relatif hampiran $\displaystyle={\left(\frac{0.0068266666666665365}{{1.65984000000000}}\right)}\cdot{100}\%={0.4113}\%$